cmd/api/src/daemons/datapipe/decoders.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2024 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package datapipe
+
+import (
+	"github.com/specterops/bloodhound/dawgs/graph"
+	"github.com/specterops/bloodhound/ein"
+	"github.com/specterops/bloodhound/log"
+	"io"
+)
+
+type ConversionFunc[T any] func(decoded T, converted *ConvertedData)
+
+func decodeBasicData[T any](batch graph.Batch, reader io.ReadSeeker, conversionFunc ConversionFunc[T]) error {
+	decoder, err := CreateIngestDecoder(reader)
+	if err != nil {
+		return err
+	}
+
+	var (
+		count         = 0
+		convertedData ConvertedData
+	)
+
+	for decoder.More() {
+		var decodeTarget T
```

Comment:
Should add a comment here as to why we are instantiating the variable in the loop's scope. A well-meaning engineer could see this as an opportunity to move the declaration into the preceding `var` statement outside of the loop.